core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +162,193 @@ public long appendEntry(JournalEntry entry) {
-      while (targetCounter > writeCounter) {
-        for (;;) {
+
+      try {
+        long startTime = System.nanoTime();
+
+        // Write pending entries to journal.
+        while (!mQueue.isEmpty()) {
           // Get, but do not remove, the head entry.
           JournalEntry entry = mQueue.peek();
           if (entry == null) {
-            // No more entries in the queue. Break out of the infinite for-loop.
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
-          writeCounter = mWriteCounter.incrementAndGet();
+          mWriteCounter++;
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if (((System.nanoTime() - startTime) >= mFlushBatchTimeNs) && !mStopFlushing) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        // Either written new entries or previous flush had been failed.
+        if (mFlushCounter.get() < mWriteCounter) {
+          mJournalWriter.flush();
+          mFlushCounter.set(mWriteCounter);
+        }
+
+        /**
+         * Notify tickets that have been served to wake up.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+              ticketIterator.remove();
             }
           }
         }
+      } catch (IOException | JournalClosedException exc) {
+        /**
+         * Release only tickets that have been flushed. Fail the rest.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            ticketIterator.remove();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+            } else {
+              ticket.setError(exc);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Submits a ticket to flush thread and waits until ticket is served.
+   *
+   * If the specified counter is already flushed, this is essentially a no-op.
+   *
+   * @param targetCounter the counter to flush
+   */
+  @SuppressWarnings("Duplicates")
```

Comment:
this should no longer be necessary, right?

Reply:
I can't even see why it was necessary. Removed it.